File: common/src/app/common/files/rebase.clj
;; This Source Code Form is subject to the terms of the Mozilla Public
;; License, v. 2.0. If a copy of the MPL was not distributed with this
;; file, You can obtain one at http://mozilla.org/MPL/2.0/.
;;
;; Copyright (c) KALEIDOS INC

(ns app.common.files.rebase
  "P2.3 — Operational Transform rebase engine.

  Given an incoming change-set (built on base-revn) and a sequence of
  competing change-sets already applied by the server (between base-revn and
  current-revn), produce a new change-set that is correctly positioned on top
  of the current server state.

  This namespace is .cljc so it can run on both the JVM backend (server-side
  rebase) and the ClojureScript frontend (optimistic client-side rebase).

  Algorithm
  ─────────
  We follow the standard two-argument transform approach from the OT
  literature (Ellis & Gibbs 1989; Ressel et al. 1996) adapted to the Logos
  change model.

  The transform matrix covers the five change types:

  │ incoming ╲ competing │ :set   │ :mod-obj │ :mov-objects │ :add-obj │ :del-obj │
  │──────────────────────┼────────┼──────────┼──────────────┼──────────┼──────────│
  │ :set (same obj+attr) │ keep*  │ keep     │ keep         │ –        │ no-op    │
  │ :set (diff attr)     │ keep   │ keep     │ keep         │ –        │ no-op    │
  │ :mod-obj             │ merge  │ merge    │ keep         │ –        │ no-op    │
  │ :mov-objects         │ keep   │ keep     │ adjust-index │ adjust   │ no-op    │
  │ :add-obj             │ keep   │ keep     │ adjust-index │ identity │ keep     │
  │ :del-obj (same id)   │ absorb │ absorb   │ absorb       │ absorb   │ identity │

  * keep = incoming change wins (represents most-recent user intent).
    This is deterministic: the server always picks the incoming change
    when two clients set the same attribute at the same time on the
    same shape.  Both clients will converge to the same state.

  No-op rules
  ───────────
  - An :add-obj that adds an id that was deleted by a competing :del-obj
    is preserved (re-add wins over delete for collaborative resilience).
  - A :del-obj cancels any competing :set or :mod-obj on the same id
    (delete takes precedence — the object is gone).
  - A :mov-objects whose target :shapes list contains ids deleted by
    competing :del-obj operations has those ids pruned from the list.

  Index adjustment for :mov-objects
  ──────────────────────────────────
  When two clients both move objects within the same parent container,
  indices can drift.  We apply a simple linear adjustment:

    If competing move removed N objects before our target index,
    adjust our index by −N.  If it inserted M objects before our index,
    adjust by +M.  This is the same heuristic used by most collaborative
    editors and gives deterministic, reasonable results for independent
    reorders.

  Limitations (acceptable for Phase 2)
  ──────────────────────────────────────
  - Nested tree conflicts (e.g., two simultaneous reparentings of the same
    subtree) are resolved as \"incoming wins\" — the last server-confirmed
    move determines final position.
  - Property-based tests (see test/common/files/rebase_test.cljc) verify
    idempotency and correctness for the common cases.

  References
  ──────────
  - Ellis & Gibbs (1989) — \"Concurrency Control in Groupware Systems\"
  - *Designing Data-Intensive Applications* ch. 5 — \"Multi-Leader Replication\"
  - Jupiter (Nichols et al. 1995) — server-authoritative OT"
  (:require [clojure.set :as set]))

;; ──────────────────────────────────────────────────────────────────
;; Helpers
;; ──────────────────────────────────────────────────────────────────

(defn- extract-set-ops
  "Return a map of {:attr → :val} for all :set operations in a :mod-obj change."
  [change]
  (->> (:operations change)
       (filter #(= :set (:type %)))
       (into {} (map (fn [op] [(:attr op) (:val op)])))))

(defn- merge-set-ops
  "Merge operations from two :mod-obj changes.
  For operations touching the same attribute, `incoming-ops` wins."
  [incoming-ops competing-ops]
  (let [incoming-attrs (set (map :attr incoming-ops))
        surviving-competing (remove #(contains? incoming-attrs (:attr %)) competing-ops)]
    (into (vec surviving-competing) incoming-ops)))

(defn- obj-ids-deleted-by
  "Return the set of object ids deleted by a single competing change."
  [competing]
  (if (= :del-obj (:type competing))
    #{(:id competing)}
    #{}))

(defn- all-deleted-ids
  "Return the set of all object ids deleted across a seq of competing change-sets."
  [competing-changes]
  (reduce (fn [acc ch]
            (into acc (obj-ids-deleted-by ch)))
          #{}
          competing-changes))

;; ──────────────────────────────────────────────────────────────────
;; Single-change rebase against one competing change
;; ──────────────────────────────────────────────────────────────────

(defmulti ^:private transform-against
  "Rebase one incoming `change` against one `competing` change.
  Returns the (possibly modified) incoming change, or nil to drop it."
  (fn [incoming competing] [(:type incoming) (:type competing)]))

;; Default: preserve incoming, no adjustment needed
(defmethod transform-against :default
  [incoming _competing]
  incoming)

;; ── :mod-obj vs :del-obj (same id) ──────────────────────────────

(defmethod transform-against [:mod-obj :del-obj]
  [incoming competing]
  (when (not= (:id incoming) (:id competing))
    incoming))

;; ── :mod-obj vs :mod-obj (same id) — merge operations ───────────

(defmethod transform-against [:mod-obj :mod-obj]
  [incoming competing]
  (if (not= (:id incoming) (:id competing))
    incoming
    ;; Merge: incoming's set-ops win over competing's for same attribute
    (let [merged-ops (merge-set-ops (:operations incoming) (:operations competing))]
      (assoc incoming :operations merged-ops))))

;; ── :del-obj vs :mod-obj (same id) — del wins ───────────────────

(defmethod transform-against [:del-obj :mod-obj]
  [incoming competing]
  ;; del-obj wins — discard the competing mod-obj implicitly (we just keep del)
  incoming)

(defmethod transform-against [:del-obj :del-obj]
  [incoming competing]
  ;; If both try to delete the same object, second one is a no-op.
  ;; Return incoming unchanged (idempotent delete is safe).
  incoming)

;; ── :add-obj vs :del-obj (same id) — add wins (re-creation) ─────

(defmethod transform-against [:add-obj :del-obj]
  [incoming competing]
  ;; The object was deleted, then the incoming change re-adds it.
  ;; Re-add wins — this is the collaborative resilience rule.
  incoming)

;; ── :mov-objects vs :del-obj — prune deleted ids from shapes list

(defmethod transform-against [:mov-objects :del-obj]
  [incoming competing]
  (let [deleted-id (:id competing)
        pruned     (into [] (remove #(= deleted-id %)) (:shapes incoming))]
    (if (empty? pruned)
      nil ;; all shapes were deleted — drop the move entirely
      (assoc incoming :shapes pruned))))

;; ── :mov-objects vs :add-obj — adjust index if insert is before target

(defmethod transform-against [:mov-objects :add-obj]
  [incoming competing]
  ;; If the competing add inserted an object into the same parent container
  ;; at an index at or before our target index, shift our index up by 1.
  (if (and (= (:parent-id incoming) (:parent-id competing))
           (some? (:index incoming))
           (some? (:index competing))
           (<= (:index competing) (:index incoming)))
    (update incoming :index inc)
    incoming))

;; ── :mov-objects vs :mov-objects — adjust index ─────────────────

(defmethod transform-against [:mov-objects :mov-objects]
  [incoming competing]
  ;; Only adjust when we're moving within the same parent container.
  (if (not= (:parent-id incoming) (:parent-id competing))
    incoming
    ;; Competing removed some shapes from this container and inserted them
    ;; elsewhere (or re-ordered them within this container).  Adjust our
    ;; target index: for each competing shape that was at an index ≤ ours
    ;; and is no longer here, decrement our index by 1; for each competing
    ;; shape that landed at an index ≤ ours from outside, increment by 1.
    ;; Simplified: if competing moved shapes out (before our index), shift down.
    (let [competing-shapes (set (:shapes competing))
          incoming-idx     (or (:index incoming) 0)
          ;; Assume competing shapes were at index < incoming / moved to index < incoming
          ;; For safety use the conservative approximation: each competing shape
          ;; that we don't also move and was before our position shifts us by 1.
          our-shapes       (set (:shapes incoming))
          interfering      (set/difference competing-shapes our-shapes)
          competing-from   (or (:index competing) 0)
          adjustment       (cond
                             ;; competing moved shapes from before our index (frees slots)
                             (< competing-from incoming-idx) (- (count interfering))
                             ;; competing inserted at or before our index (fills slots)
                             (<= (:index competing) incoming-idx) (count interfering)
                             :else 0)]
      (update incoming :index #(max 0 (+ (or % 0) adjustment))))))

;; ──────────────────────────────────────────────────────────────────
;; Public API
;; ──────────────────────────────────────────────────────────────────

(defn rebase-change
  "Rebase a single incoming `change` against all `competing-changes`
  (in the order they were applied by the server).

  Returns the rebased change, or nil if the change becomes a no-op
  (e.g., modifying a shape that was concurrently deleted)."
  [change competing-changes]
  (reduce (fn [ch competing]
            (if (nil? ch)
              (reduced nil) ;; already dropped, short-circuit
              (transform-against ch competing)))
          change
          competing-changes))

(defn rebase-change-set
  "Rebase an entire change-set (a vector of changes) against a sequence of
  `competing-change-sets` (each is itself a seq of changes).

  Each competing change-set represents one server-applied revision that was
  not yet visible to the client when it built `change-set`.

  Returns a (possibly smaller) change-set that is safe to apply on top of
  the current server state.

  The client should provide all change-sets from `base-revn` (exclusive)
  to `current-revn` (inclusive) as `competing-change-sets`.

  Example
  ───────
  Client built its change-set based on server revn 5.
  Server is now at revn 8 (three commits: 6, 7, 8).
  Competing-change-sets = [changes@revn6, changes@revn7, changes@revn8].
  rebase-change-set produces a new change-set safe to apply at revn 9."
  [change-set competing-change-sets]
  (let [;; Flatten all competing change-sets into a single ordered sequence
        ;; of individual changes for the transform matrix.
        flat-competing (into [] (mapcat identity) competing-change-sets)]
    (->> change-set
         (keep #(rebase-change % flat-competing))
         (vec))))

File: backend-go/internal/rebase/rebase.go
// Package rebase implements the P2.3 Operational Transform rebase engine.
//
// Pure-Go port of common/src/app/common/files/rebase.cljc and the Rust crate
// rust/logos-rebase/src/lib.rs.  The algorithm, transform matrix, and test
// suite all mirror those sources exactly so correctness can be verified by
// running the same test vectors in all three implementations.
//
// Algorithm summary (OT, two-argument transform):
//
//	│ incoming ╲ competing │ :mod-obj │ :del-obj │ :add-obj │ :mov-objects │
//	│──────────────────────┼──────────┼──────────┼──────────┼──────────────│
//	│ :mod-obj             │ merge    │ no-op    │ keep     │ keep         │
//	│ :del-obj             │ keep     │ identity │ keep     │ keep         │
//	│ :add-obj             │ keep     │ keep     │ keep     │ keep         │
//	│ :mov-objects         │ keep     │ prune-id │ adj-idx  │ adj-idx      │
//
// "keep"    = incoming change is returned unchanged.
// "no-op"   = incoming change is dropped (returns nil).
// "merge"   = incoming :set-ops win over competing ones for the same attr.
// "prune-id"= deleted id is removed from the :shapes list; nil if list empties.
// "adj-idx" = :index is adjusted up/down to account for concurrent inserts/removes.
//
// References:
//   - Ellis & Gibbs (1989) — "Concurrency Control in Groupware Systems"
//   - Nichols et al. (1995) — Jupiter server-authoritative OT
package rebase

import "encoding/json"

// ─── Change types ─────────────────────────────────────────────────────────────

// Type identifies the kind of shape operation in a change-set.
type Type string

const (
	TypeModObj     Type = "mod-obj"
	TypeDelObj     Type = "del-obj"
	TypeAddObj     Type = "add-obj"
	TypeMovObjects Type = "mov-objects"
)

// SetOp is a single attribute mutation inside a ModObj change.
// It mirrors {:type :set :attr "…" :val …} in the Clojure model.
// Val is left as raw JSON so no type assumptions are made about the value.
type SetOp struct {
	Attr string          `json:"attr"`
	Val  json.RawMessage `json:"val"`
}

// Change is one operation in a change-set.
//
// All fields are present in the JSON representation; the fields used by the
// rebase algorithm depend on the Type:
//
//	TypeModObj:     Type, ID, PageID, Operations
//	TypeDelObj:     Type, ID, PageID
//	TypeAddObj:     Type, ID, PageID, ParentID, Index, Obj (opaque)
//	TypeMovObjects: Type, PageID, Shapes, ParentID, Index
//
// Other types (library ops, page-level ops, etc.) pass through the transform
// matrix unchanged via the default rule.
type Change struct {
	Type       Type            `json:"type"`
	ID         string          `json:"id,omitempty"`
	PageID     string          `json:"pageId,omitempty"`
	ParentID   string          `json:"parentId,omitempty"`
	Index      int             `json:"index,omitempty"`
	Shapes     []string        `json:"shapes,omitempty"`
	Operations []SetOp         `json:"operations,omitempty"`
	Obj        json.RawMessage `json:"obj,omitempty"` // opaque add-obj payload
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

// mergeSetOps merges operations from two ModObj changes.
// For operations touching the same attribute, incomingOps wins.
// Mirrors Clojure's merge-set-ops / Rust's merge_set_ops.
func mergeSetOps(incomingOps, competingOps []SetOp) []SetOp {
	incomingAttrs := make(map[string]struct{}, len(incomingOps))
	for _, op := range incomingOps {
		incomingAttrs[op.Attr] = struct{}{}
	}

	result := make([]SetOp, 0, len(incomingOps)+len(competingOps))
	for _, op := range competingOps {
		if _, ok := incomingAttrs[op.Attr]; !ok {
			result = append(result, op)
		}
	}
	return append(result, incomingOps...)
}

// ─── Transform matrix (5×5 dispatch) ─────────────────────────────────────────

// transformAgainst rebases one incoming change against one competing change.
// Returns the (possibly modified) incoming change, or nil to drop it.
// Mirrors Clojure's defmulti transform-against and Rust's transform_against.
func transformAgainst(incoming Change, competing Change) *Change {
	switch {

	// ── :mod-obj vs :del-obj (same id) → drop incoming ─────────────────────
	case incoming.Type == TypeModObj && competing.Type == TypeDelObj:
		if incoming.ID == competing.ID {
			return nil
		}
		return &incoming

	// ── :mod-obj vs :mod-obj (same id) → merge set-ops ─────────────────────
	case incoming.Type == TypeModObj && competing.Type == TypeModObj:
		if incoming.ID != competing.ID {
			return &incoming
		}
		merged := mergeSetOps(incoming.Operations, competing.Operations)
		result := incoming
		result.Operations = merged
		return &result

	// ── :del-obj vs :del-obj (same id) → idempotent, keep ──────────────────
	case incoming.Type == TypeDelObj && competing.Type == TypeDelObj:
		// Two deletes of the same object; safe to re-delete.
		return &incoming

	// ── :add-obj vs :del-obj (same id) → re-add wins ───────────────────────
	case incoming.Type == TypeAddObj && competing.Type == TypeDelObj:
		// Collaborative resilience: re-creation takes precedence.
		return &incoming

	// ── :mov-objects vs :del-obj → prune deleted id from :shapes ───────────
	case incoming.Type == TypeMovObjects && competing.Type == TypeDelObj:
		pruned := make([]string, 0, len(incoming.Shapes))
		for _, s := range incoming.Shapes {
			if s != competing.ID {
				pruned = append(pruned, s)
			}
		}
		if len(pruned) == 0 {
			return nil // all shapes were deleted — drop the move entirely
		}
		result := incoming
		result.Shapes = pruned
		return &result

	// ── :mov-objects vs :add-obj → adjust index if insert is before target ──
	case incoming.Type == TypeMovObjects && competing.Type == TypeAddObj:
		if incoming.ParentID == competing.ParentID && competing.Index <= incoming.Index {
			result := incoming
			result.Index++
			return &result
		}
		return &incoming

	// ── :mov-objects vs :mov-objects → adjust index ─────────────────────────
	case incoming.Type == TypeMovObjects && competing.Type == TypeMovObjects:
		if incoming.ParentID != competing.ParentID {
			return &incoming
		}
		// For each competing shape that we don't also move, determine how it
		// shifts our target index (mirrors Clojure's conservative approximation).
		incomingSet := make(map[string]struct{}, len(incoming.Shapes))
		for _, s := range incoming.Shapes {
			incomingSet[s] = struct{}{}
		}
		interfering := 0
		for _, s := range competing.Shapes {
			if _, ok := incomingSet[s]; !ok {
				interfering++
			}
		}
		var adjustment int
		if competing.Index < incoming.Index {
			adjustment = -interfering // competing freed slots before our index
		} else if competing.Index <= incoming.Index {
			adjustment = interfering // competing filled slots at or before our index
		}
		result := incoming
		result.Index = max(0, incoming.Index+adjustment)
		return &result

	// ── Default: preserve incoming, no adjustment needed ────────────────────
	default:
		return &incoming
	}
}

// ─── Public API ───────────────────────────────────────────────────────────────

// RebaseChange rebases a single incoming change against all competing changes
// (in the order they were applied by the server).
//
// Returns the rebased change, or nil if the change becomes a no-op
// (e.g., modifying a shape that was concurrently deleted).
//
// Mirrors Clojure rebase-change / Rust rebase_change.
func RebaseChange(change Change, competing []Change) *Change {
	result := &change
	for _, comp := range competing {
		if result == nil {
			return nil
		}
		result = transformAgainst(*result, comp)
	}
	return result
}

// RebaseChangeSet rebases an entire change-set against a sequence of competing
// change-sets (each representing one server-applied revision the client has
// not yet seen).
//
// Returns a (possibly smaller) change-set that is safe to apply on top of
// the current server state.
//
// Example: client built its change-set based on server revn 5.
// Server is now at revn 8.  Provide competing change-sets for revn 6, 7, 8.
// The returned change-set is safe to apply at revn 9.
//
// Mirrors Clojure rebase-change-set / Rust rebase_change_set.
func RebaseChangeSet(changes []Change, competingChangeSets [][]Change) []Change {
	// Flatten all competing change-sets into a single ordered sequence.
	var flat []Change
	for _, cs := range competingChangeSets {
		flat = append(flat, cs...)
	}

	result := make([]Change, 0, len(changes))
	for _, ch := range changes {
		r := RebaseChange(ch, flat)
		if r != nil {
			result = append(result, *r)
		}
	}
	return result
}
